File: src/use-cases/party/StartParty.js
/**
 * StartParty Use Case
 * Handles starting a party game
 */

const Round = require('../../domain/entities/Round');
const GameState = require('../../domain/value-objects/GameState');
const logger = require('../../../logger');

class StartParty {
    /**
     * @param {IPartyRepository} partyRepository - Party repository
     * @param {IUserRepository} userRepository - User repository
     */
    constructor(partyRepository, userRepository) {
        this.partyRepository = partyRepository;
        this.userRepository = userRepository;
    }

    /**
     * Execute the use case
     * @param {Object} request - Start party request
     * @param {string} request.userId - User ID
     * @param {string} request.partyId - Party ID
     * @returns {Promise<Object>} Start result
     */
    async execute({ userId, partyId }) {
        try {
            // Validate input
            if (!userId || typeof userId !== 'string') {
                throw new Error('User ID is required');
            }

            if (!partyId || typeof partyId !== 'string') {
                throw new Error('Party ID is required');
            }

            // Verify user exists
            const user = await this.userRepository.findById(userId);
            if (!user) {
                throw new Error('User not found');
            }

            // Find party
            const party = await this.partyRepository.findById(partyId);
            if (!party) {
                throw new Error('Party not found');
            }

            // Check if user is the owner
            if (party.ownerId !== userId) {
                throw new Error('Only the party owner can start the game');
            }

            // Check party status
            if (party.status === 'playing') {
                throw new Error('Party is already playing');
            }

            if (party.status === 'finished') {
                throw new Error('Party has finished');
            }

            // Get players
            const players = await this.partyRepository.getPlayers(partyId);

            if (players.length < 2) {
                throw new Error('At least 2 players required to start');
            }

            // Start the party
            party.start();

            // Create first round
            const round = Round.create(partyId, 1);

            // Create initial game state
            const gameState = GameState.createInitial(players.length);

            // Initialize deck with 54 cards (52 + 2 jokers)
            const deck = Array.from({ length: 54 }, (_, i) => i);

            // Shuffle deck (Fisher-Yates)
            for (let i = deck.length - 1; i > 0; i--) {
                const j = Math.floor(Math.random() * (i + 1));
                [deck[i], deck[j]] = [deck[j], deck[i]];
            }

            // Deal cards to players
            const handSize = party.settings.handSize || 10;
            const hands = {};

            for (let i = 0; i < players.length; i++) {
                hands[i] = deck.splice(0, handSize);
            }

            // Create game state with dealt cards
            const initialGameState = gameState.with({
                deck: deck,
                hands: hands,
                currentTurn: 0,
                currentAction: 'draw',
                roundNumber: 1
            });

            // Save round
            await this.partyRepository.saveRound(round);

            // Save game state
            await this.partyRepository.saveGameState(partyId, initialGameState);

            // Update party with round ID
            party.startNewRound(round.id);
            await this.partyRepository.save(party);

            logger.info('Party started', {
                userId: userId,
                username: user.username,
                partyId: partyId,
                roundId: round.id,
                playerCount: players.length
            });

            return {
                success: true,
                party: {
                    id: party.id,
                    status: party.status,
                    currentRoundId: party.currentRoundId
                },
                round: {
                    id: round.id,
                    roundNumber: round.roundNumber,
                    status: round.status
                },
                gameState: {
                    currentTurn: initialGameState.currentTurn,
                    currentAction: initialGameState.currentAction,
                    deckSize: initialGameState.deck.length,
                    handSizes: Object.keys(hands).reduce((acc, idx) => {
                        acc[idx] = hands[idx].length;
                        return acc;
                    }, {})
                }
            };
        } catch (error) {
            logger.error('Start party error', {
                userId,
                partyId,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = StartParty;

File: zapzap-rust/src/application/party/start_party.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::domain::entities::{Party, PartyStatus, Round};
use crate::domain::repositories::{PartyRepository, RepositoryError};
use crate::domain::services::initialize_round;

/// Start party input
pub struct StartPartyInput {
    pub user_id: String,
    pub party_id: String,
}

/// Start party output
pub struct StartPartyOutput {
    pub party: Party,
    pub round: Round,
}

/// Start party use case
pub struct StartParty<P: PartyRepository> {
    party_repo: Arc<P>,
}

impl<P: PartyRepository> StartParty<P> {
    pub fn new(party_repo: Arc<P>) -> Self {
        Self { party_repo }
    }

    pub async fn execute(
        &self,
        input: StartPartyInput,
    ) -> Result<StartPartyOutput, StartPartyError> {
        // Find party
        let mut party = self
            .party_repo
            .find_by_id(&input.party_id)
            .await?
            .ok_or(StartPartyError::PartyNotFound)?;

        // Check if user is owner
        if party.owner_id != input.user_id {
            return Err(StartPartyError::NotOwner);
        }

        // Check party status
        match party.status {
            PartyStatus::Waiting => {}
            PartyStatus::Playing => return Err(StartPartyError::PartyPlaying),
            PartyStatus::Finished => return Err(StartPartyError::PartyFinished),
        }

        // Get players
        let mut players = self.party_repo.get_party_players(&input.party_id).await?;

        // Check minimum players
        if !party.can_start(players.len()) {
            return Err(StartPartyError::NotEnoughPlayers);
        }

        // The game state indexes seats 0..n-1: close the gaps a leave left. In seat
        // order, each player moves down to a seat the ones before it have just vacated.
        players.sort_by_key(|p| p.player_index);
        for (seat, player) in players.iter().enumerate() {
            if player.player_index != seat as u8 {
                self.party_repo
                    .set_player_index(&input.party_id, &player.user_id, seat as u8)
                    .await?;
            }
        }

        // Create first round
        let round_id = Uuid::new_v4().to_string();
        let round = Round::new(round_id.clone(), input.party_id.clone(), 1, 0);

        // Initialize game state
        let scores = [0u16; 8];
        let game_state = initialize_round(
            players.len() as u8,
            party.settings.hand_size,
            &scores,
            0, // No eliminated players yet
            1, // Round 1
            0, // Player 0 starts
            None,
        );

        // Save round
        self.party_repo.save_round(&round).await?;

        // Save game state
        self.party_repo
            .save_game_state(&input.party_id, &game_state)
            .await?;

        // Update party status
        party.start();
        party.current_round_id = Some(round_id);
        self.party_repo.save(&party).await?;

        Ok(StartPartyOutput { party, round })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum StartPartyError {
    #[error("Party not found")]
    PartyNotFound,
    #[error("Not the party owner")]
    NotOwner,
    #[error("Party is already playing")]
    PartyPlaying,
    #[error("Party has finished")]
    PartyFinished,
    #[error("Not enough players (minimum 3)")]
    NotEnoughPlayers,
    #[error("Repository error: {0}")]
    Repository(#[from] RepositoryError),
}
